site/documentation/content/api/event/index.md

```diff
@@ -130,3 +132,88 @@ The example below might be used by the MQTT adapter to indicate that a connectio
   }
 }
 ~~~
+
+
+## Event API for Kafka
+
+Hono can alternatively be configured to publish event messages to an Apache Kafka&reg; cluster instead of an AMQP Messaging Network. The Event API for Kafka is defined by means of Kafka messages.
+
+### Southbound Operations
+
+The following operation can be used by *Protocol Adapters* to send event messages received from devices to downstream consumers like *Business Applications*.
+
+#### Produce Event
+
+The protocol adapter writes messages to the tenant-specific topic `event-${tenant_id}` where `${tenant_id}` is the ID of the tenant that the client wants to upload event messages for.
+
+
+**Preconditions**
+
+1. Either the topic `event-${tenant_id}` exists, or the broker is configured to automatically create topics on demand.
+1. The client is authorized to write to the topic.
+1. The device for which the adapter wants to send event messages has been registered (see [Device Registration API]({{< relref "/api/device-registration" >}})).
+
+**Message Flow**
+
+Hono supports *AT LEAST ONCE* delivery of *Event* messages only. A client therefore MUST set *acks* to `all`. Either MUST *enable.idempotence* be set to `true` OR *max.in.flight.requests.per.connection* MUST be set to `1`.
+
+
+The recommended Kafka producer properties for *Event* messages are: `enable.idempotence=true`, `acks=all`, `max.in.flight.requests.per.connection=5`, leaving *retries* unset (which defaults to `Integer.MAX`), and setting *delivery.timeout.ms* to a reasonable value (supported from Kafka version 1.0.0 on). 
+
+The following sequence diagram illustrates the flow of messages involved in the *MQTT Adapter* producing an event to the Kafka cluster.
+
+{{< figure src="produce_kafka.svg" title="Produce event flow" >}}
+
+1. *Device* `4711` publishes an event using MQTT QoS 1.
+   1. *MQTT Adapter* produces an event message to *Kafka Cluster*.
+   1. *Kafka cluster* acknowledges reception of the message.
+   1. *MQTT Adapter* acknowledges the reception of the message to the *Device*.
+
+When the Kafka producer receives an exception when sending an event message to Kafka, the protocol adapter MUST NOT try to re-send such rejected messages but MUST indicate the failed transfer to the device if the transport protocol provides means to do so.
```

Comment:
Yes. The Kafka client distinguishes different types of errors. For those that are potentially transient, retries are made as configured (you can configure the maximum retry count and/or a timeout, and you can disable retries), for those that are final, it throws an exception immediately.

Reply:
@b-abel how about using the word "raises" or "throws" instead of "receives"? When I read this, it first appears as if the producer is receiving an exception (in the form of a message) from the Kafka cluster.

Reply:
maybe also replacing "when" with "while" since the error is raised during the sending process. Also, how about replacing "the" with "a" since you are referring to any Kafka producer and not a specific one.. e.g `When *a* Kafka producer *raises* an exception *while* sending an event message to Kafka...`

Reply:
@Alfusainey You are right, this is poorly formulated. I will have a look at this next week.

Reply:
@Alfusainey Thank you for your great suggestion. I applied it.

Reply:
The current specification of the [Telemetry API](https://www.eclipse.org/hono/docs/api/telemetry/) says:

> When the AMQP Messaging Network fails to settle the transfer of a telemetry message or settles the transfer with any other outcome than `accepted`, the protocol adapter MUST NOT try to re-send such rejected messages but SHOULD indicate the failed transfer to the device if the transport protocol provides means to do so.

For the event API it says: "...but MUST indicate the failed transfer...".

@sophokles73 I am not sure about the reasoning behind this and this should be applied to the Kafka based API. Why must the protocol adapter not implement a retry? And does the difference regarding the obligation to inform the device only arise from the fact that this is not possible with QoS 0?